sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ReflectionConverters.scala

```diff
@@ -0,0 +1,263 @@
+        val iter = structType.fields.iterator
+        var idx = 0
+        while (iter.hasNext) {
+          converters(idx) = createCatalystConverter(iter.next().dataType)
+          idx += 1
+        }
+        (item: Any) => {
+          extractOption(item) match {
+            case r: Row => convertRowWithConverters(r, structType, converters)
+
+            case p: Product =>
+              val ar = new Array[Any](structType.size)
+              val iter = p.productIterator
+              var idx = 0
+              while (idx < structType.size) {
+                ar(idx) = converters(idx)(iter.next())
+                idx += 1
+              }
+              new GenericRowWithSchema(ar, structType)
+
+            case null => null
+          }
+        }
+
+      case _ =>
+        (item: Any) => extractOption(item) match {
+          case d: BigDecimal => Decimal(d)
+          case d: java.math.BigDecimal => Decimal(d)
+          case d: java.sql.Date => DateUtils.fromJavaDate(d)
+          case other => other
+        }
+    }
+  }
+
+  /** Converts Catalyst types used internally in rows to standard Scala types */
+  def convertToScala(a: Any, dataType: DataType): Any = (a, dataType) match {
+    // Check UDT first since UDTs can override other types
+    case (d, udt: UserDefinedType[_]) => udt.deserialize(d)
+    case (s: Seq[_], arrayType: ArrayType) => s.map(convertToScala(_, arrayType.elementType))
+    case (m: Map[_, _], mapType: MapType) => m.map { case (k, v) =>
+      convertToScala(k, mapType.keyType) -> convertToScala(v, mapType.valueType)
+    }
+    case (r: Row, s: StructType) => convertRowToScala(r, s)
+    case (d: Decimal, _: DecimalType) => d.toJavaBigDecimal
+    case (i: Int, DateType) => DateUtils.toJavaDate(i)
+    case (other, _) => other
+  }
+
+  /**
+   * Creates a converter function that will convert Catalyst types to Scala type.
+   */
+  private[sql] def createScalaConverter(dataType: DataType): Any => Any = dataType match {
+    // Check UDT first since UDTs can override other types
+    case udt: UserDefinedType[_] =>
+      (item: Any) => if (item == null) null else udt.deserialize(item)
+
+    case arrayType: ArrayType =>
+      val elementConverter = createScalaConverter(arrayType.elementType)
+      (item: Any) => if (item == null) null else item.asInstanceOf[Seq[_]].map(elementConverter)
+
+    case mapType: MapType =>
+      val keyConverter = createScalaConverter(mapType.keyType)
+      val valueConverter = createScalaConverter(mapType.valueType)
+      (item: Any) => if (item == null) {
+        null
+      } else {
+        item.asInstanceOf[Map[_, _]].map { case (k, v) =>
+          keyConverter(k) -> valueConverter(v)
+        }
+      }
+
+    case s: StructType =>
+      val converters = createScalaConvertersForStruct(s)
+      (item: Any) => item match {
+        case r: Row => convertRowWithConverters(r, s, converters)
+        case other => other
```

Comment:
The one for row wasn't, but for DateType, for example, I think it makes sense to leave other clause.